File: design-patterns/golang/facade/deployment_facade.go
package facade

import "fmt"

// DeploymentFacade provides a simplified interface to the deployment system
type DeploymentFacade struct {
	vcs              *VersionControlSystem
	buildSystem      *BuildSystem
	testingFramework *TestingFramework
	deploymentTarget *DeploymentTarget
}

// NewDeploymentFacade creates a new DeploymentFacade instance
func NewDeploymentFacade() *DeploymentFacade {
	return &DeploymentFacade{
		vcs:              NewVersionControlSystem(),
		buildSystem:      NewBuildSystem(),
		testingFramework: NewTestingFramework(),
		deploymentTarget: NewDeploymentTarget(),
	}
}

// DeployApplication performs a full standard deployment
func (f *DeploymentFacade) DeployApplication(branch, serverAddress string) bool {
	fmt.Printf("\nFACADE: --- Initiating FULL DEPLOYMENT for branch: %s to %s ---\n", branch, serverAddress)
	success := true

	// Step 1: Pull latest code
	f.vcs.PullLatestChanges(branch)

	// Step 2: Build the project
	if !f.buildSystem.CompileProject() {
		fmt.Println("FACADE: DEPLOYMENT FAILED - Build compilation failed.")
		return false
	}
	artifactPath := f.buildSystem.GetArtifactPath()

	// Step 3: Run tests
	if !f.testingFramework.RunUnitTests() {
		fmt.Println("FACADE: DEPLOYMENT FAILED - Unit tests failed.")
		return false
	}
	if !f.testingFramework.RunIntegrationTests() {
		fmt.Println("FACADE: DEPLOYMENT FAILED - Integration tests failed.")
		return false
	}

	// Step 4: Deploy to production
	f.deploymentTarget.TransferArtifact(artifactPath, serverAddress)
	f.deploymentTarget.ActivateNewVersion(serverAddress)

	fmt.Printf("FACADE: APPLICATION DEPLOYED SUCCESSFULLY TO %s!\n", serverAddress)
	return success
}

// DeployHotfix performs a hotfix deployment
func (f *DeploymentFacade) DeployHotfix(branch, serverAddress string) bool {
	fmt.Printf("\nFACADE: --- Initiating HOTFIX DEPLOYMENT for branch: %s to %s ---\n", branch, serverAddress)
	success := true

	// Step 1: Pull latest code
	f.vcs.PullLatestChanges(branch)

	// Step 2: Build the project
	if !f.buildSystem.CompileProject() {
		fmt.Println("FACADE: HOTFIX FAILED - Build compilation failed.")
		return false
	}
	artifactPath := f.buildSystem.GetArtifactPath()

	// Step 3: Skip extensive tests for hotfix
	fmt.Println("FACADE: Skipping full test suite for hotfix deployment (or running minimal smoke tests).")

	// Step 4: Deploy to production
	f.deploymentTarget.TransferArtifact(artifactPath, serverAddress)
	f.deploymentTarget.ActivateNewVersion(serverAddress)

	fmt.Printf("FACADE: HOTFIX DEPLOYED SUCCESSFULLY TO %s!\n", serverAddress)
	return success
}

File: design-patterns/cpp/facade/deployment_facade.cpp
#include "deployment_facade.h"
#include <iostream>

DeploymentFacade::DeploymentFacade()
    : vcs(std::make_unique<VersionControlSystem>())
    , buildSystem(std::make_unique<BuildSystem>())
    , testingFramework(std::make_unique<TestingFramework>())
    , deploymentTarget(std::make_unique<DeploymentTarget>()) {}

bool DeploymentFacade::deployApplication(const std::string& branch, const std::string& serverAddress) {
    std::cout << "\nFACADE: --- Initiating FULL DEPLOYMENT for branch: " << branch << " to " << serverAddress << " ---" << std::endl;
    bool success = true;

    try {
        // Step 1: Pull latest code
        vcs->pullLatestChanges(branch);

        // Step 2: Build the project
        if (!buildSystem->compileProject()) {
            std::cerr << "FACADE: DEPLOYMENT FAILED - Build compilation failed." << std::endl;
            return false;
        }
        std::string artifactPath = buildSystem->getArtifactPath();

        // Step 3: Run tests
        if (!testingFramework->runUnitTests()) {
            std::cerr << "FACADE: DEPLOYMENT FAILED - Unit tests failed." << std::endl;
            return false;
        }
        if (!testingFramework->runIntegrationTests()) {
            std::cerr << "FACADE: DEPLOYMENT FAILED - Integration tests failed." << std::endl;
            return false;
        }

        // Step 4: Deploy to production
        deploymentTarget->transferArtifact(artifactPath, serverAddress);
        deploymentTarget->activateNewVersion(serverAddress);

        std::cout << "FACADE: APPLICATION DEPLOYED SUCCESSFULLY TO " << serverAddress << "!" << std::endl;
    } catch (const std::exception& e) {
        std::cerr << "FACADE: DEPLOYMENT FAILED - An unexpected error occurred: " << e.what() << std::endl;
        success = false;
    }
    return success;
}

bool DeploymentFacade::deployHotfix(const std::string& branch, const std::string& serverAddress) {
    std::cout << "\nFACADE: --- Initiating HOTFIX DEPLOYMENT for branch: " << branch << " to " << serverAddress << " ---" << std::endl;
    bool success = true;

    try {
        // Step 1: Pull latest code
        vcs->pullLatestChanges(branch);

        // Step 2: Build the project
        if (!buildSystem->compileProject()) {
            std::cerr << "FACADE: HOTFIX FAILED - Build compilation failed." << std::endl;
            return false;
        }
        std::string artifactPath = buildSystem->getArtifactPath();

        // Step 3: For a hotfix, we skip extensive tests
        std::cout << "FACADE: Skipping full test suite for hotfix deployment (or running minimal smoke tests)." << std::endl;

        // Step 4: Deploy to production
        deploymentTarget->transferArtifact(artifactPath, serverAddress);
        deploymentTarget->activateNewVersion(serverAddress);

        std::cout << "FACADE: HOTFIX DEPLOYED SUCCESSFULLY TO " << serverAddress << "!" << std::endl;
    } catch (const std::exception& e) {
        std::cerr << "FACADE: HOTFIX FAILED - An unexpected error occurred: " << e.what() << std::endl;
        success = false;
    }
    return success;
} 